[relay-runtime] Use 'rawResponse' instead of 'response' for optimisticResponse

janicduplessis opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/relay/blob/86842c141eb59b31269893f224bf24375a2b3539/packages/relay-runtime/mutations/commitMutation.js#L71
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

The type for `optimisticResponse` should use the generated `rawResponse` value instead of `response`. `rawResponse` will include fields added or changed by the relay compiler and that are expected in the response. This is documented here https://relay.dev/docs/guided-tour/updating-data/graphql-mutations/#optimistic-response.

Note this is a breaking change since it is possible the generated type is missing if the mutation is not annotated with `@raw_response_type`. In some cases it will still typecheck fine, but could cause bugs if added fields like `id` are missing from the optimistic response object. To migrate annotate the mutation with `@raw_response_type`.

rbuckton commented:
I don't feel like I understand this change well enough to make an informed decision as to whether to merge this change. It would help if a maintainer with a deeper understanding weighed in.

@alloy @maraisr @edvinerikson @Stephen2 @morrys, are you able to review this?

Stephen2 reviewed (approved):
I had the perfect code to test this.  It's a good improvement.

Unfortunately it doesn't fail if:
- I leave off `@raw_response_type`
- I pass rubbish code:

```
      optimisticResponse: {
        z: "shouldFail",
      },
```

But as soon as I add `@raw_response_type` it forced me to add things like `id` which had previously not been enforced at compile time.

Fixing the failure case seems way out of scope of this PR.

Great stuff mate

janicduplessis commented:
@Stephen2 Turns out the type of the field becomes `unknown` when `rawResponse` is not generated. I added a check and made it `never` instead so it doesn't typecheck when passing random values.

I also added the documentation from the relay website so it will appear inline that you need to add `@raw_response_type` to the mutatation since it is not obvious from the TS error.

Stephen2 reviewed (approved):
clever

orta commented:
Looks reasonable to me 👍🏻

jquense commented:
Can someone provide some more context for this change. We just upgraded and have all of our usages of `optimisticResponse` break, now complaining that they should be `undefined`, when previously they worked fine. I see references here to a directive we've never used `@raw_response_type` which is basically undocumented except for a passing note about types. Would anyone be able to share what we should do to fix this and why this is necessary vs the previous behavior which seemed to work correctly already?

jquense commented:
testing this out we're seeing the "raw response" generate types that do not seem correct for mutation responses with Unions in them. in particular it's requiring we provide fields for every possible field across the union, instead of just one of the concrete types
